media-proxy/include/mesh/st2110rx.h

```diff
@@ -0,0 +1,174 @@
+        ops.name = strdup(session_name);
+        ops.framebuff_cnt = 4;
+
+        ops.priv = this; // app handle register to lib
+        ops.notify_frame_available = frame_available_cb;
+
+        log::info("ST2110Rx: configure")("port", ops.port.port[MTL_PORT_P])(
+            "ip_addr", std::to_string(ops.port.ip_addr[MTL_PORT_P][0]) + " " +
+                           std::to_string(ops.port.ip_addr[MTL_PORT_P][1]) + " " +
+                           std::to_string(ops.port.ip_addr[MTL_PORT_P][2]) + " " +
+                           std::to_string(ops.port.ip_addr[MTL_PORT_P][3]))(
+            "mcast_sip_addr", std::to_string(ops.port.mcast_sip_addr[MTL_PORT_P][0]) + " " +
+                                  std::to_string(ops.port.mcast_sip_addr[MTL_PORT_P][1]) + " " +
+                                  std::to_string(ops.port.mcast_sip_addr[MTL_PORT_P][2]) + " " +
+                                  std::to_string(ops.port.mcast_sip_addr[MTL_PORT_P][3]))(
+            "num_port", ops.port.num_port)("udp_port", ops.port.udp_port[MTL_PORT_P])(
+            "name", ops.name)("framebuff_cnt", ops.framebuff_cnt);
+
+        return 0;
+    }
+
+    Result on_establish(context::Context& ctx) override {
+        _ctx = context::WithCancel(ctx);
+        stop = false;
+
+        mtl_session = create_session(mtl_device, &ops);
+        if (!mtl_session) {
+            log::error("Failed to create session");
+            set_state(ctx, State::closed);
+            return set_result(Result::error_general_failure);
+        }
+
+        /* Start MTL session thread. */
+        try {
+            frame_thread_handle = std::jthread(&ST2110Rx::frame_thread, this);
+        } catch (const std::system_error& e) {
+            log::error("Failed to create thread");
+            set_state(ctx, State::closed);
+            return set_result(Result::error_out_of_memory);
+        }
+
+        set_state(ctx, State::active);
+        return set_result(Result::success);
+    }
+
+    Result on_shutdown(context::Context& ctx) override {
+        _ctx.cancel();
+        stop = true;
+        stop.notify_all();
+
+        frame_thread_handle.join();
+
+        if (mtl_session) {
+            close_session(mtl_session);
+            mtl_session = nullptr;
+        }
+        set_state(ctx, State::closed);
+        return set_result(Result::success);
+    };
+
+    virtual void on_delete(context::Context& ctx) override {}
+
+  private:
+    void frame_thread() {
+        while (!_ctx.cancelled()) {
+            // Get full buffer from MTL
+            FRAME *frame_ptr = get_frame(mtl_session);
+            if (!frame_ptr) { /* no frame */
+                stop.wait(false);
+                stop = false;
+                continue;
+            }
+            // Forward buffer to emulated receiver
+            transmit(_ctx, get_frame_data_ptr(frame_ptr), transfer_size);
+            // Return used buffer to MTL
+            put_frame(mtl_session, frame_ptr);
```

Comment:
```suggestion
            if (frame_ptr) {
                // Forward buffer to emulated receiver
                transmit(_ctx, get_frame_data_ptr(frame_ptr), transfer_size);
                // Return used buffer to MTL
                put_frame(mtl_session, frame_ptr);
            } else {
                frame_available.wait(true, std::memory_order_acquire);
            }
```

Reply:
Fixed in 133a6efcdea8895252fd1a825a2ae57a7f31cb01